File: src/main/java/Main/Inventory.java
package Main;

import com.fasterxml.jackson.annotation.JsonIgnore;
import Main.ItemClasses.Consumable;
import Main.ItemClasses.Item;
import Main.ItemClasses.Miscellaneous;
import Main.ItemClasses.Weapons.Weapon;
import kotlin.Pair;

import java.util.ArrayList;
import java.util.UUID;

public class Inventory {
    private ArrayList<Item> items = new ArrayList<>();
    private final ArrayList<Spell> spells = new ArrayList<>();
    private String name = "Inventory";
    private volatile ArrayList<Integer> spellSlotsUsed = new ArrayList<>();
    private volatile ArrayList<Integer> spellSlotsMax = new ArrayList<>();
    private boolean loadedLevels = false;
    private float maxCarryingCapacity = 100f;
    @JsonIgnore
    public UUID uuid = UUID.randomUUID();
    @JsonIgnore
    private final ArrayList<Pair<Integer, Integer>> spellLevels = new ArrayList<>();


    public Inventory() {}

    public Inventory(String name) {
        this.name = name;
        addLastSpellLevel(new Pair<>(3, 3));
    }

    public Inventory(String name, UUID uuid, ArrayList<Item> items) {
        this.name = name;
        this.uuid = uuid;
        this.items = items;
    }

    @JsonIgnore
    public Inventory(Inventory other) {
        this.name = other.name;
        this.uuid = UUID.randomUUID();
        this.items = new ArrayList<>(other.items);
        this.spells.addAll(other.spells);
        this.spellSlotsUsed.addAll(other.spellSlotsUsed);
        this.spellSlotsMax.addAll(other.spellSlotsMax);
        this.loadedLevels = other.loadedLevels;
        this.maxCarryingCapacity = other.maxCarryingCapacity;
        this.spellLevels.addAll(other.spellLevels);
    }

    public String getName() {
        return this.name;
    }

    public ArrayList<Item> getItems() {
        return items;
    }

    public float getMaxCarryingCapacity() {
        return maxCarryingCapacity;
    }

    public void setMaxCarryingCapacity(float maxCarryingCapacity) {
        this.maxCarryingCapacity = maxCarryingCapacity;
    }

    public ArrayList<Spell> getSpells() {
        return spells;
    }

    public void addLastSpellLevel(Pair<Integer, Integer> level) {
        spellLevels.addLast(level);
    }

    public void removeSpellLevel(int index) {
        spellLevels.remove(index);
    }

    public ArrayList<Pair<Integer, Integer>> getSpellLevels() {
        if(!loadedLevels) {
            loadedLevels = true;
            spellSlotsUsed.forEach(used ->
                    spellLevels.add(new Pair<>(used, spellSlotsMax.get(spellSlotsUsed.indexOf(used))))
            );
        }

        System.out.println("spellSlots: " + spellLevels);
        System.out.println("spellSlotsUsed: " + spellSlotsUsed);
        System.out.println("spellSlotsMax: " + spellSlotsMax);
        return spellLevels;
    }

    public void resetUsedSpellSlots() {
        spellSlotsUsed.forEach(used -> spellSlotsUsed.set(spellSlotsUsed.indexOf(used), spellSlotsMax.get(spellSlotsUsed.indexOf(used))));
        for(Pair<Integer, Integer> level : spellLevels) {
            int index = spellLevels.indexOf(level);
            int a = level.component1();
            int b = level.component2();
            Pair<Integer, Integer> replacement = new Pair<>(b, b);
            spellLevels.set(index, replacement);
        }
    }
}

File: src/main/kotlin/main/Inventory.kt
package main

import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.snapshots.SnapshotStateList
import main.ItemClasses.Item
import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient
import main.CharacterManager.selectedInventory
import main.ItemClasses.Armor
import main.ItemClasses.Consumable
import main.ItemClasses.EmptySlot
import main.ItemClasses.Miscellaneous
import main.ItemClasses.Potion
import main.ItemClasses.Weapons.LongRangeWeapon
import main.ItemClasses.Weapons.ShortRangeWeapon
import java.util.*
import kotlin.collections.ArrayList

@Serializable
class Inventory(
    var name: String = "Inventory",
    private var items: MutableList<Item> = mutableListOf(),
    val spells: ArrayList<Spell> = ArrayList(),
    var spellSlotsUsed: ArrayList<Int> = ArrayList(),
    var spellSlotsMax: ArrayList<Int> = ArrayList(),
    var maxCarryingCapacity: Float = 100f
) {
    val uuid: String = UUID.randomUUID().toString()
    val totalSlots = 50

    @Transient
    private var loadedLevels = false
    
    @Transient
    private val _spellLevels = ArrayList<Pair<Int, Int>>()

    init {
        // If this is a truly new inventory (not being deserialized)
        if (spellSlotsUsed.isEmpty() && spellSlotsMax.isEmpty() && items.isEmpty() && spells.isEmpty()) {
            addLastSpellLevel(3 to 3)
        }
    }

    constructor(other: Inventory) : this(
        name = other.name,
        items = ArrayList(other.items),
        spells = ArrayList(other.spells),
        spellSlotsUsed = ArrayList(other.spellSlotsUsed),
        spellSlotsMax = ArrayList(other.spellSlotsMax),
        maxCarryingCapacity = other.maxCarryingCapacity
    )

    fun addLastSpellLevel(level: Pair<Int, Int>) {
        spellSlotsUsed.add(level.first)
        spellSlotsMax.add(level.second)
        if (loadedLevels) {
            _spellLevels.add(level)
        }
    }

    fun removeSpellLevel(index: Int) {
        if (index in spellSlotsUsed.indices) {
            spellSlotsUsed.removeAt(index)
            spellSlotsMax.removeAt(index)
            if (loadedLevels && index in _spellLevels.indices) {
                _spellLevels.removeAt(index)
            }
        }
    }

    fun getSpellLevels(): ArrayList<Pair<Int, Int>> {
        if (!loadedLevels) {
            loadedLevels = true
            _spellLevels.clear()
            for (i in spellSlotsUsed.indices) {
                if (i < spellSlotsMax.size) {
                    _spellLevels.add(Pair(spellSlotsUsed[i], spellSlotsMax[i]))
                }
            }
        }
        return _spellLevels
    }

    fun resetUsedSpellSlots() {
        for (i in spellSlotsMax.indices) {
            spellSlotsUsed[i] = spellSlotsMax[i]
        }
        if (loadedLevels) {
            for (i in _spellLevels.indices) {
                val current = _spellLevels[i]
                _spellLevels[i] = Pair(current.second, current.second)
            }
        }
    }

    val typePriority = mapOf(
        ShortRangeWeapon::class to 0,
        LongRangeWeapon::class to 1,
        Armor::class to 2,
        Potion::class to 3,
        Consumable::class to 4,
        Miscellaneous::class to 5
    )

    fun getItems(): MutableList<Item> {
        return (items.take(totalSlots).toMutableList() + List(totalSlots - items.size) { EmptySlot() }.toMutableList()) as MutableList<Item>
    }

    fun getItemsSortedByClass(): MutableList<Item> {
        val processedItems = items.sortedWith(compareBy { item -> typePriority[item::class] ?: Int.MAX_VALUE })
        return (processedItems.take(totalSlots).toMutableList() + List(totalSlots - processedItems.size) { EmptySlot() }.toMutableList()) as MutableList<Item>
    }

    fun addItem(item: Item) {
        items.add(item)
    }

    fun removeItem(item: Item) {
        items.remove(item)
    }

    //if index is -1 add last
    fun addItemAtIndex(item: Item, hoveredItem: Item) {
        if(hoveredItem is EmptySlot) {
            items.addLast(item)
        }
        else {
            val dropIndex = items.indexOf(hoveredItem)
            if (dropIndex != -1) {
                items.add(dropIndex, item)
            } else {
                items.addLast(item)
            }
        }
    }

    fun updateSpellSlotsFromLevels() {
        if (loadedLevels) {
            spellSlotsUsed.clear()
            spellSlotsMax.clear()
            for (level in _spellLevels) {
                spellSlotsUsed.add(level.first)
                spellSlotsMax.add(level.second)
            }
        }
    }
}
